contents/blog/temporal-exports.md

```diff
@@ -0,0 +1,131 @@
+
+### 2. Opaque
+
+On top of unreliability, we didn’t have a clear picture of what was happening in the export apps. Throughout our codebase, we had multiple different work queues like Celery, Kafka, and Graphile each running separately in different contexts without full visibility into those contexts.
+
+We didn’t have logs or reports for export apps. For example, we didn’t know how much data was being exported or when duplicate runs were happening. When issues happened, we relied on users to tell us. This led to a lot of firefighting and often manual restarts of exports.
+
+### 3. Expensive
+
+The final nail was that export apps were expensive. This wasn’t because we were charging a lot for exports – in fact, we weren’t charging anything. It was because they were expensive on the destination side. 
+
+The unreliability led to larger amounts of data processing on the destination side, all of which come with costs. For example, if your export failed 75% through and you needed to reset, you end up paying ingestion fees of 175%. We were also paying for more processing than needed because exports were failing repeatedly.
+
+On top of extra processing time, export apps were taking up large amounts of our team’s time. Reliability needed fixing before we worked on performance, and firefighting always focused on reliability. More importantly, our team was spending large amounts of time firefighting and "babysitting" exports. They had to be reactive, rather than proactive.
+
+Simply, we outgrew the export apps and needed a more reliable, transparent, and cost-effective system if we wanted to scale further.
+
+## Temporal to the rescue
+
+Early in 2023, one of our engineers, [James](/community/profiles/90), opened an [RFC recommending using Temporal at PostHog](https://github.com/PostHog/meta/pull/99). Temporal is a workflow engine abstracting away the details of failure modes, retry logic, and timeouts. This enables developers to build and deploy rock-solid business logic. It is used by big companies with large amounts of data flows – like Uber, Coinbase, Doordash, and Hashicorp.
+
+In Q2, we set a goal of "rock solid-batch processing" and [chose Temporal](https://github.com/PostHog/meta/pull/99) to build it because it helped solve issues like:
+
+- Retry, resume, cancel, and timeout logic per activity and workflow.
+- Provides assurances about running exactly-once.
+- Visibility into the status of workflows, including errors and logging.
+- Scheduling. 
+
+Instead of building all this functionality ourselves, Temporal takes care of it for us. After some debate, we felt building a workflow engine would not be a competitive advantage for us. Nothing about what we build on Temporal is novel enough to justify an investment into a competing runtime (as satisfying as it would be to write).
```

Comment:
```suggestion
Instead of building all this functionality ourselves, Temporal takes care of it for us. After some debate, we felt building a workflow engine would not be a competitive advantage for us. As satisfying as it would be to write, nothing about what we build on Temporal is novel enough to justify an investment into a competing runtime.
```